### PR TITLE
[API-54] Occurrence Reports

### PR DESCRIPTION
**Pull request by quetzyg**
## Description
This pull request addresses issue [API-54](https://github.com/vostpt/api/issues/54).

## Task items:
- [X] Added `Occurrence` report endpoints: generate and download;
- [X] Updated the API documentation with the new endpoints;
- [X] Added test coverage for the new endpoints;
- [X] Added `failed_jobs` migration;
- [X] Added `Report` class + interface;
- [X] Added `OccurrenceReport` class;
- [X] Added `ReportGenerator` job class;

## Requirements
- Migrations need to be executed (`php artisan migrate`);
- New Composer dependencies must be installed (`composer install`);
- Supervisor must be configured to run jobs (see [documentation](https://laravel.com/docs/5.8/queues#supervisor-configuration));

### CODE REVIEW

File: app/Reports/OccurrenceReport.php [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+use Illuminate\Database\Query\JoinClause;
+
+class OccurrenceReport extends Report
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function getHeader(): array
+    {
+        return \array_merge([
```
Comment (josepostiga) [inline app/Reports/OccurrenceReport.php]: Why the call to `array_merge()`?

File: app/Http/Requests/Occurrence/DownloadReport.php [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+declare(strict_types=1);
+
+namespace VOSTPT\Http\Requests\Occurrence;
+
+use VOSTPT\Http\Requests\Request;
+
+class DownloadReport extends Request
```
[the remark this answers is not in the document]
Comment (quetzyg) [reply app/Http/Requests/Occurrence/DownloadReport.php]: In effect, yes. But I like to have a request class for each endpoint, regardless if there are no rules or authorize logic.

File: app/Http/Requests/Occurrence/GenerateReport.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            ],
+            'order' => [
+                Rule::in(OccurrenceFilter::getOrderValues()),
+            ],
+        ];
+
+        if ($this->has('started_at', 'ended_at')) {
+            $rules = \array_merge_recursive($rules, [
+                'started_at' => [
+                    'date_format:Y-m-d',
```
[the remark this answers is not in the document]
Comment (quetzyg) [reply app/Http/Requests/Occurrence/GenerateReport.php]: You are correct!

File: app/Jobs/Report/ReportGenerator.php [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    /**
+     * Execute the job.
+     *
+     * @param \Psr\Log\LoggerInterface $logger
+     *
+     * @return bool
+     */
+    public function handle(LoggerInterface $logger): bool
+    {
+        return $this->report->isReadyForDownload() || $this->report->generate($logger);
```
[the remark this answers is not in the document]
Comment (quetzyg) [reply app/Jobs/Report/ReportGenerator.php]: The `generate()` method will at best return `false` if the `fileCanBeOverwritten()` method call fails, but we're also interested in the other outcome. That's why we're explicitly calling it.

File: app/Reports/OccurrenceReport.php (hunk 1 above)
Comment (quetzyg) [reply app/Reports/OccurrenceReport.php]: That is a very good question...

**Pull request merged**